lxd/db/openfga.go

```diff
@@ -0,0 +1,610 @@
+//
+// Implementation:
+//   - Since the relation is not `member`, `project`, or `server` it is an entitlement that may form part of a
+//     permission, and only groups have permissions. So we first get the ID of the entity via the URL that is
+//     part of the Object. Then we get the permission for that entity ID, entity type, and relation (entitlement).
+//     Finally, we return all groups that have that permission.
+//   - One exception to the above is the "type bound public access" (https://openfga.dev/docs/modeling/public-access)
+//     that is defined for `server` `can_view`, which allows all identities access to `GET /1.0` and `GET /1.0/storage`.
+//     We check for this case before making any DB queries.
+//
+// Notes:
+//   - The exception for the type-bound public access may be better placed as a contextual tuple. However, adding it here
+//     means we can avoid an unnecessary transaction that will happen a lot.
+//   - We will need to modify this exception when we implement service accounts.
+func (o *openfgaStore) ReadUsersetTuples(ctx context.Context, store string, filter storage.ReadUsersetTuplesFilter) (storage.TupleIterator, error) {
+	// Expect both an object and a relation.
+	if filter.Object == "" || filter.Relation == "" {
+		return nil, fmt.Errorf("ReadUsersetTuples: Filter must include both an object and a relation")
+	}
+
+	// Expect a URL to be present for the object. (E.g. we don't want to list all groups that have `can_view` on
+	// all projects, we should be checking for a specific entity).
+	entityTypeStr, entityURL, hasURL := strings.Cut(filter.Object, ":")
+	if !hasURL {
+		return nil, fmt.Errorf("ReadUsersetTuples: Listing all entities of type not supported")
+	}
+
+	entityType := entity.Type(entityTypeStr)
+	err := entityType.Validate()
+	if err != nil {
+		return nil, fmt.Errorf("ReadUsersetTuples: Invalid object filter %q: %w", filter.Object, err)
+	}
+
+	// Check for type-bound public access exception.
+	if entityType == entity.TypeServer && filter.Relation == "can_view" {
+		return storage.NewStaticTupleIterator([]*openfgav1.Tuple{
+			// Only returning one tuple here for the identity. When adding service accounts, we'll need
+			// to add another tuple to account for them.
+			{
+				Key: &openfgav1.TupleKey{
+					Object:   fmt.Sprintf("%s:%s", entity.TypeServer, entity.ServerURL().String()),
+					Relation: string(auth.EntitlementCanView),
+					User:     fmt.Sprintf("%s:*", entity.TypeIdentity),
+				},
+			},
+		}), nil
+	}
+
+	u, err := url.Parse(entityURL)
+	if err != nil {
+		return nil, fmt.Errorf("ReadUsersetTuples: Failed to parse entity URL %q: %w", entityURL, err)
+	}
+
+	var groupNames []string
+	err = o.clusterDB.Transaction(ctx, func(ctx context.Context, tx *ClusterTx) error {
+		// Get the ID of the entity.
+		entityRef, err := cluster.GetEntityReferenceFromURL(ctx, tx.Tx(), &api.URL{URL: *u})
+		if err != nil {
+			return err
+		}
+
+		// Get all groups with the permission.
+		q := `
```

Comment:
As part of the refactoring tasks please can this query be moved into the clusterDB package. Thanks